### ISSUE TITLE
syntax highlight in html

### ISSUE THREAD

**Issue by jahagirdar** [vimwiki/vimwiki#621]
Is it possible for the exported html to have syntax highlighting for the embedded code fragment?

**Comment by optik-aper** [comment]
Not native to vimwiki, as far as I know.  You could include something like [highlight.js](https://highlightjs.org/), I'd think, but I've never used it.

**Comment by gpanders** [comment]
You can do this as follows:

1. Copy the file `default.tpl` out of `vimwiki/autoload/vimwiki/` into your own directory.
2. Modify your copy of `default.tpl` to add the following lines:

In `<head>`:
```html
<link rel="stylesheet" type="text/css" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.15.6/styles/default.min.css" />
```

At the end of `<body>`:
```html
<script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.15.6/highlight.min.js"></script>
<script type="text/javascript">
    document.querySelectorAll('pre').forEach(block => hljs.highlightBlock(block));
</script>
```
Now tell Vimwiki that you want to use this template:
```vim
let g:vimwiki_list = [{'path': '/path/to/wikis', 'template_path': '/path/to/directory/containing/default.tpl'}]
```

Highlight.js will detect syntax automatically. So in your Vimwiki files, you just need to use the `{{{` and `}}}` blocks to designate a `<pre>` section and highlight.js will take care of the rest.

You can also "hint" at the language to help highlight.js by using `{{{class="python"` (or whatever the language you are using is) in your Vimwiki file.

**Comment by hq6** [comment]
I am a new maintainer, and I think this would be nice to have natively. 

However, I will defer to @EinfachToll  and @Nudin until I get a better sense for what flavor of changes they think are appropriate, so you may want to hold off on this for now, while we handle the current PR's.

**Comment by Nudin** [comment]
Hey, I really like to have syntax-highlithing in the HTML-export. But I also think we shouldn't add JS to the default template since most users won't need it. And I think we should handle this and the math-rendering consistently.
Currently for the math rendering we explain in the help-file how to enable it. So we could do the same for this. On the other side I fear that the documentation if becoming to long so that people might stop reading it. I'm planing to create a wiki where to collect tips and tricks – documenting the how-to there and just referencing it from the help-file might be the second option.
The third option would be to have an option for both (highlight.js and mathjax and potentially more in the future) that can infuse the code to the templates and add a variable to the templates where to add these codes. So that you can enable them by a single boolean option. But I'm not sure about that yet. Any comments?

**Comment by gpanders** [comment]
@Nudin I think setting it in an option (default `0`) is a fine idea. This would require only minimal documentation additions and would work out of the box for most users. Anyone who wants to tweak the default syntax highlighting/MathJAX options has the ability to do that by using a custom template.

**Comment by optik-aper** [comment]
I completely agree with @gpanders that this shouldn't be a default.  I personally do everything I can to avoid a dependency on javascript and it's easy enough to add in as needed.  When I initially looked at handling the syntax highlighting without highlight.js I decided to try rendering color markup for the code snippets through a python library (the name escapes me now) but I haven't actually sat down to get it working end to end and so I didn't suggest it.  The .js solution is probably the easiest anyway.

**Comment by HugoNikanor** [comment]
A while ago I looked at integrating highlight(1) with VimWiki. I prefer that solution since it generates HTML tags for the code highlighting during the build step, not relying on javascript.

The problem I found with adding the feature was that the current HTML creator doesn't keep track of what's in the code block, but rather just creates start and end tags for them (I believe).

**Comment by jahagirdar** [comment]
For a non JS implementation how easy/difficult would it be to pass the codeblock through vim's builtin `:TOhtml` command?

**Comment by brennen** [comment]
This is probably a reasonable approach, but it would definitely need to be behind some configuration; we can't assume the external dependencies.  Also keep in mind that it shouldn't be OS-dependent.

**Comment by tinmarino** [comment]
Fixed by #1181 1181
